### PR TITLE
Report total time at max active threads

### PR DESCRIPTION
**Pull request by al97**
These metrics are needed as a part of an internal project for thread scaling.
Total time spent at max active threads will be joined with worker cpu utilization to see if we are bounded by the # of parallel work items, ultimately to reduce worker idleness.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by jrmccluskey** [comment]
assign to next reviewer

**Comment by jrmccluskey** [comment]
You can run `./gradlew :runners:google-cloud-dataflow-java:worker:spotlessApply` from the top level of the directory to fix the formatting issues

**Comment by jrmccluskey** [comment]
Still working on this?

**Comment by al97** [comment]
assign to next reviewer

**Comment by jrmccluskey** [comment]
@bvolpato do you mind giving this a look? Java is not in my wheelhouse

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java [hunk 1]
```diff
@@ -51,7 +55,31 @@ public BoundedQueueExecutor(
             keepAliveTime,
             unit,
             new LinkedBlockingQueue<>(),
-            threadFactory);
+            threadFactory) {
+          @Override
+          protected void beforeExecute(Thread t, Runnable r) {
+            super.beforeExecute(t, r);
+            synchronized (this) {
+              if (activeCount.get() == maximumPoolSize - 1) {
```
Comment (bvolpato) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:63]: nit I'm always afraid of off-by-one errors, but I guess a `>=` would be preferred.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java [hunk 2]
```diff
@@ -51,7 +55,31 @@ public BoundedQueueExecutor(
...
+                startTimeMaxActiveThreadsUsed = System.currentTimeMillis();
+              }
+              activeCount.incrementAndGet();
```
Comment (bvolpato) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:66]: We could make it "more" atomic

```suggestion
              if (activeCount.getAndIncrement() == maximumPoolSize - 1) {
                startTimeMaxActiveThreadsUsed = System.currentTimeMillis();
              }
```

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java [hunk 3]
```diff
@@ -51,7 +55,31 @@ public BoundedQueueExecutor(
...
+            }
+          }
+
+          @Override
+          protected void afterExecute(Runnable r, Throwable t) {
+            super.afterExecute(r, t);
+            synchronized (this) {
+              if (activeCount.get() == maximumPoolSize) {
```
Comment (bvolpato) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:74]: nit Same idea, getAndDecrement()... I guess you'd just need the synchronized block on the long

**Review by bvolpato — APPROVED** [review]
This LGTM, and seems like a very useful metric! Thanks

I've left a few comments / suggestions, but all minor

**Comment by bvolpato** [comment]
Run Java PreCommit

**Review by bvolpato — APPROVED** [review]

**Comment by jrmccluskey** [comment]
Once the last workflows pass I will merge. Thank you!

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java [hunk 4]
```diff
@@ -2852,6 +2853,74 @@ public void testActiveWorkForShardedKeys() throws Exception {
     Mockito.verifyNoMoreInteractions(mockExecutor);
   }
 
+  @Test
+  public void testMaxThreadMetric() throws Exception {
```
Comment (mosche) [inline runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:2857]: Hi @al97 
would you mind having another look at this test? It looks flaky on Jenkins (#27555).
Thanks so much :)

**Pull request merged**